keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,902 @@
+   in the `DeviceCounterConsumption` field on the `ResourceSliceMixins`.
+
+With these changes, attributes, capacity, and counters that are shared across
+devices or counter sets can be split out into mixins, thereby reducing
+duplication and reducing the size of the ResourceSlice object.
+
+### Risks and Mitigations
+
+This change doesn't really affect the functionality of DRA, it just
+provides a more compact way to define devices in ResourceSlices. But it
+still have some consequences worth pointing out here.
+
+#### ResourceSlice resources will be harder to understand
+
+The biggest challenge with this change is that it adds a level of
+indirection for the `Device` and `CounterSet` definitions, meaning
+that it gets harder to understand the ResourceSlice objects.
+
+We have discussed adding a kubectl command or a plugin that will allow
+users to see the fully flattened versions of a ResourceSlice. But this
+is not in scope for alpha.
+
+#### More attributes, capacities and counters might worsen worst-case scheduling
+
+With mixins, DRA driver authors can choose to create more complex devices,
+which might lead to worse scheduling performance. But it is up to DRA driver
+authors to do this, so it is not something that can be triggered by normal users.
+
+## Design Details
+
+### API
+
+The exact set of proposed API changes can be seen below:
+```go
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+  ...
+
+  // Mixins defines the mixins available for devices and counter sets
+  // in the ResourceSlice.
+  //
+  // +featureGate=DRAResourceSliceMixins
+  // +optional
+  Mixins *ResourceSliceMixins
+}
+
+type CounterSet struct {
+  ...
+
+  // Includes defines a list of references to CounterSetMixin.
+  // The counters listed in the mixins will be added to the counters
+  // available in this CounterSet.
+  //
+  // The counters of each included mixin are applied to this counter set in
+  // order. Conflicting counters from multiple mixins are taken from the
+  // last mixin listed. Counters set on the CounterSet will always override
+  // counters from mixins.
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of includes is 8.
+  //
+  // +featureGate=DRAResourceSliceMixins
+  // +listType=atomic
+  // +optional
+  Includes []CounterSetMixinRef
+}
+
+// CounterSetMixinRef defines a reference to a CounterSetMixin.
+type CounterSetMixinRef struct {
+  // Name refers to a CounterSetMixin defined in the same
+  // ResourceSlice.
+  //
+  // +required
+  Name string
+}
+
+// ResourceSliceMixins defines mixins for the ResourceSlice.
+//
+// The main purposes of these mixins is to reduce the memory footprint
+// of devices since they can reference the mixins provided here rather
+// than duplicate them.
+type ResourceSliceMixins struct {
```

Comment:
Sorry if this initiates a lengthy, annoying naming debate! But: because this is a struct would it be preferable to call this `ResourceSliceMixinSpec` instead (and then we'd ostensibly update the changes to `ResourceSliceSpec` to include a new `MixinSpec` property (instead of a `Mixins` property).

Reply:
Interesting question, but when I look at the existing fields on the `ResourceSliceSpec`, they don't use the `...Spec` naming convention in the field name or the type. So for consistency, I think we should just follow the pattern and keep the current names.